how to go from bootstrap (1.5.1) to 1.15.1

## Bug Report

This is about building latest crystal on FreeBSD, in
the ports & binary pkg system. I'm one of the ports
committers, making this work for all users of FreeBSD.

- FreeBSD builds everything from sources "because"
- ✅ thus we need to bootstrap from LLVM14 -> Crystal 1.5.1
- ✅ then we can build successfully LLVM16 -> Crystal 1.10.1
- ✅ then we can build successfully LLVM20 -> Crystal 1.15.1

But skipping the intermediary Crystal 1.10.1 fails, i.e.
 🤯  trying to go directly from bootstrap -> LLVM20 -> Crystal 1.15.1 fails:

```
===>  Building for crystal-1.15.1
Using /usr/local/bin/llvm-config19 [version=19.1.7]
CRYSTAL_CONFIG_BUILD_COMMIT="84f389ac5424" CRYSTAL_CONFIG_PATH=lib:/usr/local/lib/crystal SOURCE_DATE_EPOCH="1738627200"  CRYSTAL_CONFIG_LIBRARY_PATH='$ORIGIN/../lib/crystal' ./bin/crystal build --progress --stats --threads 16 --release --no-debug -D strict_multi_assign -D preview_overload_order --link-flags="  -fstack-protector-strong "  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib -D use_pcre2
Using compiled compiler at .build/crystal
Parse:                             00:00:00.000026540 (   1.08MB)
Semantic (top level):              00:00:00.457428870 ( 186.13MB)
Semantic (new):                    00:00:00.002755877 ( 186.13MB)
Semantic (type declarations):      00:00:00.054514598 ( 194.13MB)
Semantic (abstract def check):     00:00:00.018375190 ( 218.13MB)
Semantic (restrictions augmenter): 00:00:00.020660960 ( 218.13MB)
Showing last frame. Use --error-trace for full trace.

There was a problem expanding macro 'macro_35976034720'

Code in src/gc/boehm.cr:428:3

 428 | {% if flag?(:preview_mt) %}
       ^
Called macro defined in src/gc/boehm.cr:428:3

 428 | {% if flag?(:preview_mt) %}

Which expanded to:

 > 2 | # support for legacy gc releases
 > 3 | def self.set_stackbottom(stack_bottom : Void*)
 > 4 |   LibGC.stackbottom = stack_bottom
               ^----------
Error: undefined fun 'stackbottom=' for LibGC
```

From my perspective its better to only need 1 bootstrap step rather than 2.

Is the above bug fixable?

Or should I go the less satisfying but workable multi-stage bootstrap via Crystal 1.10.1?

thanks!

thanks! [the patch](https://git.sr.ht/~dch/ports/commit/d694ff9a41f397caa96b95186493dc9746c6ed23) gets us all the way through the build but craps out during linker,
I should be able to fix this here.

```
=======================<phase: build          >============================
===== env: DEVELOPER_MODE=yes STRICT_DEPENDS=yes USER=dch UID=1002 GID=0
===>  Building for crystal-1.15.1
Using /usr/local/bin/llvm-config20 [version=20.1.2]CRYSTAL_CONFIG_BUILD_COMMIT="84f389ac5424" CRYSTAL_CONFIG_PATH=lib:/usr/local/lib/crystal SOURCE_DATE_EPOCH="1738627200"  CRYSTAL_CONFIG_LIBRARY_PATH='$ORIGIN/../lib/crystal' ./bin/crystal build --progress --stats --threads 1 --release --no-debug -D strict_multi_assign -D preview_overload_order --link-flags="  -fstack-protector-strong "  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib -D use_pcre2
Using compiled compiler at .build/crystal
[1/13] Parse                             
Parse:                             00:00:00.000028684 (   1.08MB)
...  
Semantic (cvars initializers):     00:00:00.008975492 (1938.74MB)
ld: error: unable to find library -lgc-threaded [33;1m(this usually means you need to install the development package for libgc-threaded)

cc: error: linker command failed with exit code 1 (use -v to see invocation)

Error: execution of command failed with code: 1: `cc "${@}" -o /wrkdirs/usr/ports/lang/crystal/work/cache/wrkdirs-usr-ports-lang-crystal-work-crystal-1.15.1-src-ecr-process.cr/macro_run  -fstack-protector-strong  -rdynamic -lgc-threaded -lpthread -lpthread`
[8/13] Semantic (cvars initializers)     
[9/13] Semantic (main)                   
gmake: *** [Makefile:246: .build/crystal] Error 1
*** Error code 1
```

BTW full [log](https://pkg.skunkwerks.at/poudriere/data/current_x64-default/2025-04-04_12h40m40s/logs/errors/crystal-1.15.1.log) here.

--- a/src/gc/boehm.cr
+++ b/src/gc/boehm.cr
@@ -431,21 +431,21 @@ module GC
       sb.mem_base = stack_bottom
       LibGC.set_stackbottom(thread_handle, pointerof(sb))
     end
-  {% elsif LibGC.has_method?(:set_stackbottom) %}
-    # this is necessary because Boehm GC does _not_ use `GC_stackbottom` on
-    # Windows when pushing all threads' stacks; it also started crashing on
-    # Linux with libgc after v8.2.4; instead `GC_set_stackbottom` must be used
-    # to associate the new bottom with the running thread
-    def self.set_stackbottom(stack_bottom : Void*)
-      sb = LibGC::StackBase.new
-      sb.mem_base = stack_bottom
-      # `nil` represents the current thread (i.e. the only one)
-      LibGC.set_stackbottom(nil, pointerof(sb))
-    end
   {% else %}
-    # support for legacy gc releases
     def self.set_stackbottom(stack_bottom : Void*)
-      LibGC.stackbottom = stack_bottom
+      \{% if LibGC.has_method?(:set_stackbottom) %}
+        # this is necessary because Boehm GC does _not_ use `GC_stackbottom` on
+        # Windows when pushing all threads' stacks; it also started crashing on
+        # Linux with libgc after v8.2.4; instead `GC_set_stackbottom` must be used
+        # to associate the new bottom with the running thread
+        sb = LibGC::StackBase.new
+        sb.mem_base = stack_bottom
+        # `nil` represents the current thread (i.e. the only one)
+        LibGC.set_stackbottom(nil, pointerof(sb))
+      \{% else %}
+        # support for legacy gc releases
+        LibGC.stackbottom = stack_bottom
+      \{% end %}
     end
   {% end %}
 

